src/Features/Core/Completion/CompletionListProvider.cs

```diff
@@ -1,28 +1,27 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Text;
 
-namespace Microsoft.CodeAnalysis.Completion.Providers
+namespace Microsoft.CodeAnalysis.Completion
 {
     /// <summary>
     /// Extensibility interface for clients who want to participate in completion inside an editor.
     /// </summary>
-    internal interface ICompletionProvider
+    internal abstract class CompletionListProvider
     {
         /// <summary>
-        /// Returns a CompletionItemGroup for the specified position in the document.
+        /// Implement to register the items and other details for a <see cref="CompletionList"/>
         /// </summary>
-        Task<CompletionItemGroup> GetGroupAsync(Document document, int position, CompletionTriggerInfo triggerInfo, CancellationToken cancellationToken = default(CancellationToken));
+        public abstract Task RegisterCompletionListAsync(CompletionListContext context);
```

Comment:
I get that it sounds weird. Again, I'm just trying to match our existing extensibility pattern ATM with this internal API. Before it goes public, we'll get to polish the naming properly.